### DISCUSSION TITLE
[css-loader] `getLocalIdent` generate different class for stylesheet and for selector

### DISCUSSION THREAD

**Discussion by dawidurbanski** [webpack/webpack#17867] [Q&A]
I'm building an app that will be embedded in another software. I'm using CSS modules for styling.

I want to preserve some nice class name in order for other people to be able to adjust styling of the app from external styles.

What I'm doing now:

CSS loader config:

```js
options: {
    modules: {
        localIdentName: "[name]__[local]--[hash:base64:5]",
    },
},
```

My `style.module.css`:

```css
.hello {
    color: red;
}
```

compiles to:

```css
.Foo-module__hello--PN2l7 {
    color: red;
}
```

In my JSX (`Foo.jsx`):

```jsx
<div className={styles.hello}></div>
```

compiles to:

```html
<div class="Foo-module__hello--PN2l7"></div>
```

In general I would like `css-loader` to do this:

```css
.hello {
    color: red;
}
```

compiles to (no changes):

```css
.Foo-module__hello--PN2l7 {
    color: red;
}
```

but my HTML:

```jsx
<div className={styles.hello}></div>
```

compiles to:

```html
<div class="Foo-module__hello Foo-module__hello--PN2l7"></div>
```

_____

Is there any way of knowing in the `getLocalIdent` callback if we are generating class in CSS output stylesheet versus HTML class attribute?

Pseudocode:

```js
options: {
    modules: {
        getLocalIdent: (context, localIdentName, localName, options) => {
            if (context: we are generating stylesheet class) {
                return something like "[name]__[local]--[hash:base64:5]"
            }

            if (context: we are generating html class attribute) {
                return something like "[name]__[local] [name]__[local]--[hash:base64:5]"
            }
        },
    },
},
```

_____

I know I could just add the original class in my output JSX component like:

```jsx
<div className={`Foo-module__hello ${styles.hello}`}></div>
```

but it's annoying to do it for every `className` even if I would extract it to separate helper function.

_____

For now here are my options:

1. Add original class name in JSX (not ideal)
2. Let people know that they can style these components with `[class^=Foo-module__hello]`, again not great experience
3. Just use `[name]__[local]` without hash for class names and make sure all components names are unique to avoid conflicts.

I'm leaning towards the last one for now if what I'm trying to do is impossible.

**Comment by azangru** [comment]
> In my JSX (Foo.jsx):
> `<div className={styles.hello}></div>`
> compiles to:
> `<div class="Foo-module__hello--PN2l7"></div>`

Unless I am missing something, `css-loader` does not handle the JSX at all. Its responsibility is limited to only load CSS files; and if the `modules` options is enabled, then the CSS classes in the output are renamed, and the `import` statement returns an object mapping original class names to the transformed ones. CSS loader does not know how you use this class name mapping after it has been imported into your javascript file.

Personally, I have used your option 2; but also, tried to treat components styled with CSS modules as if they were proper web components with shadow DOM, and expose the properties that are allowed to be styled via CSS variables. I think option 3 breaks the very purpose of CSS modules. If you choose option 3, you could just as well have used a naming convention like BEM, which makes class name collisions unlikely, but does not give any guarantees.
